src/Helpers/JWKFetcher.php

```diff
@@ -59,101 +62,135 @@ protected function convertCertToPem($cert)
-     * @param string      $path   Path to the JWKS from the $domain.
+     * @throws \Exception
      *
-     * @return mixed
      *
-     * @throws \Exception If the Guzzle HTTP client cannot complete the request.
+     * @codeCoverageIgnore
      */
-    public function fetchKeys($domain, $kid = null, $path = '.well-known/jwks.json')
+    public function fetchKeys($iss)
     {
-        $jwks_url = $domain.$path;
-
-        // Check for a cached JWKS value.
-        $secret = $this->cache->get($jwks_url);
-        if (! is_null($secret)) {
-            return $secret;
-        }
-
-        $secret = [];
-
-        $jwks = $this->getJwks( $domain, $path );
-
-        if (! is_array( $jwks['keys'] ) || empty( $jwks['keys'] )) {
-            return $secret;
-        }
-
-        // No kid passed so get the kid of the first JWK.
-        if (is_null($kid)) {
-            $kid = $this->getProp( $jwks, 'kid' );
-        }
-
-        $x5c = $this->getProp( $jwks, 'x5c', $kid );
+        $url = "{$iss}.well-known/jwks.json";
+        if (($secret = $this->cache->get($url)) === null) {
+            $secret  = [];
+            $request = new RequestBuilder([
+                'domain' => $iss,
+                'basePath' => '.well-known/jwks.json',
+                'method' => 'GET',
+                'guzzleOptions' => $this->guzzleOptions
+            ]);
+            $jwks    = $request->call();
+            foreach ($jwks['keys'] as $key) {
+                $secret[$key['kid']] = $this->convertCertToPem($key['x5c'][0]);
+            }
 
-        // Need the kid and x5c for a well-formed return value. 
-        if (! is_null($kid) && ! is_null($x5c)) {
-            $secret[$kid] = $this->convertCertToPem($x5c);
-            $this->cache->set($jwks_url, $secret);
+            $this->cache->set($url, $secret);
         }
 
         return $secret;
     }
+    // phpcs:enable
 
     /**
-     * Get a specific property from a JWKS using a key, if provided.
+     * Fetch x509 cert for RS256 token decoding.
      *
-     * @param array       $jwks JWKS to parse.
-     * @param string      $prop Property to retrieve.
-     * @param null|string $kid  Kid to check.
+     * @param string      $jwks_url URL to the JWKS.
+     * @param string|null $kid      Key ID.
      *
-     * @return null|string
+     * @return string|null
      */
-    public function getProp(array $jwks, $prop, $kid = null)
+    public function getJwksX5c($jwks_url, $kid = null)
     {
-        $r_key = null;
-        if (! $kid) {
-            // No kid indicated, get the first entry.
-            $r_key = $jwks['keys'][0];
-        } else {
-            // Iterate through the JWKS for the correct kid.
-            foreach ($jwks['keys'] as $key) {
-                if (isset($key['kid']) && $key['kid'] === $kid) {
-                    $r_key = $key;
-                    break;
-                }
-            }
+        $cache_key = $jwks_url.'|'.(string) $kid;
```

Comment:
think of the case that kid is null (not passed as arg). This will also break the cache impl below ;) I personally would skip cache in this particular case

Reply:
`null` is technically a value we can handle (first key) so why not cache for that value?

Reply:
if `null` get's converted to `"null"` (note the quotes) then I guess it's a valid string and you will end up having a valid key, so yes, you will be able to handle just fine null kids (pun intended)